core/escrow/escrow.go

```diff
@@ -67,33 +68,79 @@ func NewContractRequest(configuration *config.Config, signedContracts []*escrowp
 	}, nil
 }
 
-func SubmitContractToEscrow(configuration *config.Config, request *escrowpb.EscrowContractRequest) error {
+func SubmitContractToEscrow(configuration *config.Config, request *escrowpb.EscrowContractRequest) (*escrowpb.SignedSubmitContractResult, error) {
 	var conn *grpc.ClientConn
 	// TODO: Make escrow IP hidden in config too, now for testing purpose leave it here
 	conn, err := grpc.Dial("52.15.101.94:50051", grpc.WithInsecure())
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer conn.Close()
 	client := escrowpb.NewEscrowServiceClient(conn)
 	response, err := client.SubmitContracts(context.Background(), request)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	if response == nil {
-		return fmt.Errorf("escrow reponse is nil")
+		return nil, fmt.Errorf("escrow reponse is nil")
 	}
+	// verify
+	err = verifyEscrowRes(configuration, response.Result, response.EscrowSignature)
+	if err != nil {
+		return nil, fmt.Errorf("verify escrow failed %v", err)
+	}
+	return response, nil
+}
+
+func verifyEscrowRes(configuration *config.Config, message proto.Message, sig []byte) error {
 	escrowPubkey, err := crypto.ToPubKey(configuration.Services.EscrowPubKeys[0])
 	if err != nil {
 		return err
 	}
-	ok, err := crypto.Verify(escrowPubkey, response.Result, response.EscrowSignature)
+	ok, err := crypto.Verify(escrowPubkey, message, sig)
 	if err != nil || !ok {
 		return fmt.Errorf("verify escrow failed %v", err)
 	}
 	return nil
 }
 
+func NewPayinRequest(result *escrowpb.SignedSubmitContractResult, payerPubKey ic.PubKey, payerPrivKey ic.PrivKey) (*escrowpb.SignedPayinRquest, error) {
+	chanState := result.Result.BuyerChannelState
+	sig, err := crypto.Sign(payerPrivKey, chanState.Channel)
+	if err != nil {
+		return nil, err
+	}
+	chanState.FromSignature = sig
+	payerAddr, err := payerPubKey.Raw()
+	if err != nil {
+		return nil, err
+	}
+	payinReq := &escrowpb.PayinRquest{
+		PayinId:           result.Result.PayinId,
+		BuyerAddress:      payerAddr,
+		BuyerChannelState: chanState,
+	}
+	payinSig, err := crypto.Sign(payerPrivKey, payinReq)
+	return &escrowpb.SignedPayinRquest{
+		Request:        payinReq,
+		BuyerSignature: payinSig,
+	}, nil
+}
+
+func PayInToEscrow(configuration *config.Config, signedPayinReq *escrowpb.SignedPayinRquest) error {
+	conn, err := grpc.Dial("52.15.101.94:50051", grpc.WithInsecure())
```

Comment:
@laipogo please help Shelly with this grpc conn since this is a new path of code

Reply:
> @laipogo please help Shelly with this grpc conn since this is a new path of code

ok, will do